core/lib/Cardano/Address/Script.hs

```diff
@@ -266,35 +300,75 @@ foldScript fn zero = \case
+validateScript :: Script KeyHash -> Either ErrValidateScript ()
+validateScript script = do
+    let validateKeyHash (KeyHash bytes) =
+            when (BS.length bytes /= credentialHashSize) $ Left WrongKeyHash
+    validateScriptWith validateKeyHash script
+
+validateScriptWith
+    :: Eq elem
+    => (elem -> Either ErrValidateScript ())
+    -> Script elem
+    -> Either ErrValidateScript ()
+validateScriptWith validateRequireSignatureOf = \case
+    RequireSignatureOf element ->
+        validateRequireSignatureOf element
 
     RequireAllOf script -> do
-        when (L.null script) $ Left EmptyList
+        when (L.null (omitTimelocks script)) $ Left EmptyList
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (invalidTimelocks script) $ Left InvalidTimelocks
+        traverse_ (validateScriptWith validateRequireSignatureOf) script
 
     RequireAnyOf script -> do
-        when (L.null script) $ Left EmptyList
+        when (L.null (omitTimelocks script)) $ Left EmptyList
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (invalidTimelocks script) $ Left InvalidTimelocks
+        traverse_ (validateScriptWith validateRequireSignatureOf) script
 
     RequireSomeOf m script -> do
         when (m == 0) $ Left MZero
-        when (length script < fromIntegral m) $ Left ListTooSmall
+        when (length (omitTimelocks script) < fromIntegral m) $ Left ListTooSmall
         when (hasDuplicate script) $ Left DuplicateSignatures
-        traverse_ validateScript script
+        when (invalidTimelocks script) $ Left InvalidTimelocks
+        traverse_ (validateScriptWith validateRequireSignatureOf) script
 
     ActiveFromSlot _ -> pure ()
 
     ActiveUntilSlot _ -> pure ()
   where
-    hasDuplicate xs = do
+    hasDuplicate xs =
         length sigs /= length (L.nub sigs)
       where
         sigs = [ sig | RequireSignatureOf sig <- xs ]
+    hasTimelocks = \case
+        ActiveFromSlot _ -> True
+        ActiveUntilSlot _ -> True
+        _ -> False
+    invalidTimelocks xs = case filter hasTimelocks xs of
+        [] -> False
+        [ActiveFromSlot s1, ActiveUntilSlot s2] -> s2 <= s1
+        [ActiveUntilSlot s2, ActiveFromSlot s1] -> s2 <= s1
+        [ActiveFromSlot _] -> False
+        [ActiveUntilSlot _] -> False
+        _ -> True
+    omitTimelocks = filter (not . hasTimelocks)
+--
+-- ScriptTemplate validation
+--
+
+-- | Validate a 'ScriptTemplate', semantically
+--
+-- @since 3.2.0
+validateScriptTemplate :: ScriptTemplate -> Either ErrValidateScript ()
+validateScriptTemplate (ScriptTemplate cosigners' script) = do
+    when (Map.size cosigners' == 0) $ Left NoCosigner
+    when (L.length (L.nub $ Map.elems cosigners') /= Map.size cosigners') $
+        Left DuplicateXPubs
+    let validateCosigner cosigner =
+            when (cosigner `notElem` (Map.keys cosigners')) $ Left UnknownCosigner
```

Comment:
I think we need to check inclusion both ways: 

- All cosigners referred to in the cosigners map are in the script
- All cosigners mentioned in the script are also in the cosigners map

So that we enforce a clear 1:1 mapping.

Reply:
done